server/src/main/java/org/opensearch/indices/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,241 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.indices;
+
+import org.roaringbitmap.RoaringBitmap;
+
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class RBMIntKeyLookupStore implements KeyLookupStore<Integer> {
+    // This class shares a lot of the same fields with HybridIntKeyLookupStore, but basically none of the logic
+    // besides getters, so I decided against making it the superclass to HybridIntKeyLookupStore
+    protected final int modulo;
+    protected int size;
+    protected long memSizeCapInBytes;
+    protected int numAddAttempts;
+    protected int numCollisions;
+    protected boolean guaranteesNoFalseNegatives;
+    protected RoaringBitmap rbm;
+    protected final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    protected final Lock readLock = lock.readLock();
+    protected final Lock writeLock = lock.writeLock();
+    protected RBMSizeEstimator sizeEstimator;
+    protected int maxNumEntries;
+    protected boolean atCapacity;
+
+    RBMIntKeyLookupStore(int modulo, long memSizeCapInBytes) {
+        this.modulo = modulo;
+        this.size = 0;
+        this.numAddAttempts = 0;
+        this.numCollisions = 0;
+        this.guaranteesNoFalseNegatives = true;
+        this.memSizeCapInBytes = memSizeCapInBytes; // A cap of 0 means no cap
+        this.sizeEstimator = new RBMSizeEstimator(modulo);
+        this.maxNumEntries = calculateMaxNumEntries();
+        this.rbm = new RoaringBitmap();
+    }
+
+    protected int calculateMaxNumEntries() {
+        if (memSizeCapInBytes == 0) {
+            return Integer.MAX_VALUE;
+        }
+        return sizeEstimator.getNumEntriesFromSizeInMB(RBMSizeEstimator.convertBytesToMB(memSizeCapInBytes));
+    }
+
+    protected final int transform(int value) {
+        return modulo == 0 ? value : value % modulo;
+    }
+
+    protected void handleCollisions(int transformedValue) {
+        numCollisions++;
+    }
+
+    @Override
+    public boolean add(Integer value) throws Exception {
+        if (value == null) {
+            return false;
+        }
+        writeLock.lock();
+        numAddAttempts++;
+        try {
+            if (size == maxNumEntries) {
+                atCapacity = true;
+                return false;
+            }
+            int transformedValue = transform(value);
+            boolean alreadyContained = contains(transformedValue);
+            if (!alreadyContained) {
+                rbm.add(transformedValue);
+                size++;
+                return true;
+            }
+            handleCollisions(transformedValue);
+            return false;
```

Comment:
I guess we can make this function add as void itself, as client calling add() might not find this value useful.